libraries/AP_RCTelemetry/AP_CRSF_Telem.cpp

```diff
@@ -79,20 +80,140 @@ void AP_CRSF_Telem::setup_wfq_scheduler(void)
     add_scheduler_entry(1300, 500); // battery           2Hz
     add_scheduler_entry(550, 280);  // GPS               3Hz
     add_scheduler_entry(550, 500);  // flight mode       2Hz
+    add_scheduler_entry(5000, 50);  // passthrough       max 20Hz
+    add_scheduler_entry(5000, 500); // status text       max 2Hz
+}
+
+void AP_CRSF_Telem::setup_custom_telemetry() {
+    if (_custom_telem_init_done) {
+        return;
+    }
+
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+
+    // we need crossfire firmware version
+    if (_crsf_version.pending) {
+        return; 
+    }
+
+    AP_Frsky_SPort_Passthrough* passthrough = AP::frsky_passthrough_telem();
+    if (passthrough == nullptr) {
+        return;
+    }
+    
+    // setup the frsky scheduler for crossfire
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LAT);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::GPS_LON);
+    passthrough->disable_scheduler_entry(AP_Frsky_SPort_Passthrough::TEXT);
+    passthrough->set_scheduler_entry_min_period(AP_Frsky_SPort_Passthrough::ATTITUDE, 200); // 5Hz
+
+    // setup the crossfire scheduler for custom telemetry
+    set_scheduler_entry(BATTERY, 1000, 1000);       // 1Hz
+    set_scheduler_entry(ATTITUDE, 1000, 1000);      // 1Hz
+    set_scheduler_entry(FLIGHT_MODE, 1200, 2000);   // 0.5Hz
+    set_scheduler_entry(HEARTBEAT, 2000, 5000);     // 0.2Hz
+    
+    _telem_rf_mode = get_rf_mode();
+    // setup custom telemetry for current rf_mode
+    update_custom_telemetry_rates(_telem_rf_mode);
+
+    gcs().send_text(MAV_SEVERITY_DEBUG,"CRSF: custom telemetry initialized");
+    _custom_telem_init_done = true;
+}
+
+void AP_CRSF_Telem::update_custom_telemetry_rates(AP_RCProtocol_CRSF::RFMode rf_mode) {
+    if (is_high_speed_telemetry(rf_mode)) {
+        // custom telemetry for high data rates
+        set_scheduler_entry(GPS, 550, 280);           // 3Hz
+        set_scheduler_entry(PASSTHROUGH, 50, 50);     // 20Hz
+        set_scheduler_entry(STATUS_TEXT, 100, 500);   // 2Hz
+    } else {
+        // custom telemetry for low data rates
+        set_scheduler_entry(GPS, 550, 1000);              // 1Hz
+        set_scheduler_entry(PASSTHROUGH, 500, 3000);      // 0.5Hz
+        set_scheduler_entry(STATUS_TEXT, 600, 2000);      // 0.5Hz
+    }
+}
+
+void AP_CRSF_Telem::process_rf_mode_changes()
+{
+    if (!rc().crsf_custom_telemetry()) {
+        return;
+    }
+    const AP_RCProtocol_CRSF::RFMode current_rf_mode = get_rf_mode();
+    if ( _telem_rf_mode != current_rf_mode) {
+        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d", (uint8_t)_telem_rf_mode, (uint8_t)current_rf_mode);
```

Comment:
ok, added latest packet rate info like "CRSF: rf mode change %d->%d, rate is %dHz"

Reply:
Cool. So because of the flipping that people see I think you need to introduce some latency into this reporting to avoid annoying spam. For instance only print the message if the mode has changed and stay changed for longer than the equivalent RC timeout (e.g. 200ms)

Reply:
yes, this would help, will do!

Reply:
...mhmm not sure anymore, tested it and even with a 500ms grace period is still flipping between rf mode 1 and 2, I start to believe that is correct to let this pop to the user

Reply:
this is the code I'm using
```
    const AP_RCProtocol_CRSF::RFMode current_rf_mode = get_rf_mode();
    const uint32_t now_ms = AP_HAL::millis();
    if ( _telem_rf_mode.mode != current_rf_mode && !_telem_rf_mode.change_notify_pending) {
        hal.console->printf("set: mode %d, time %lu\n", (uint8_t)_telem_rf_mode.mode, now_ms);
        _telem_rf_mode.last_change_time_ms = now_ms;
        _telem_rf_mode.change_notify_pending = true;
    } else if ((now_ms - _telem_rf_mode.last_change_time_ms) > 500 && _telem_rf_mode.change_notify_pending) {
        hal.console->printf("reset: mode %d, time %lu\n", (uint8_t)_telem_rf_mode.mode, now_ms);
        gcs().send_text(MAV_SEVERITY_INFO, "CRSF: rf mode change %d->%d, rate is %dHz", (uint8_t)_telem_rf_mode.mode, (uint8_t)current_rf_mode, _scheduler.avg_packet_rate);
        update_custom_telemetry_rates(current_rf_mode);
        _telem_rf_mode.mode = current_rf_mode;
        _telem_rf_mode.change_notify_pending = false;
    }
```

Reply:
Yeah seems like quite a big delay, so maybe leave out until people complain! How come the rate is largely unchanged fomr mode 1 <-> mode 2?

Reply:
...it doesn't change much because it's an average over 1sec, once it really switches to mode 1 it goes down (and that's the reason I did not print it before...it takes a while to settle on the real value)

Reply:
I'll park this in a "stand by PR" :-)